Caioh

caiohamamura opened the pull request:
Hi @jefferis, as I posted previously, I just added ANN directory to `/inst/include` so that ANN.h can be imported and used by other libraries.

To link to your package one shoud:
1. Add LinkingTo: RANN2 and Depends: RANN2 to the DESCRIPTION file
1. Add the PKG_LIBS to your package in Makevars and Makevars.win:
```
PKG_IMPORT=RANN2
PKG_LIBS=-L"$(R_HOME)/library/$(PKG_IMPORT)/libs$(R_ARCH)" -L"$(R_LIBS_USER)/$(PKG_IMPORT)/libs$(R_ARCH)" -l$(PKG_IMPORT)
```

You are ready to include <ANN/ANN.h> in any project using ANN.

Commit:
Allow to import ANN.h

Commit:
Added relazed unload for releasing dynamic library

jefferis commented on R/zz.R:
```diff
@@ -0,0 +1,3 @@
+.onUnload <- function (libpath) {
+  library.dynam.unload("RANN2", libpath)
```
I think it's normal to finish the function definition with `invisible()`. See results of https://github.com/search?q=.onUnLoad+user%3Acran+path%3AR&type=Code

jefferis reviewed (changes requested):
Thanks a lot for the PR @caiohamamura! 

* I would prefer not to duplicate the header files. Is it possible just to have them in inst? You may need put them in an ANN subdir.
* I made another comment on unload
* could you rename the file `R/zz.R` to `R/zzz.R`
* Could you merge in the latest master branch (I pushed some updates) and
  add yourself as a contributor (`role="ctb"`) in `Authors@R` field in DESCRIPTION

Commit:
Add Caio Hamamura as contributor
Add .vscode to gitignore
Rename zz.R t zzz.R
Make on.Unload return invisible()
Remove ANN headers duplication
Fix Makevars include path to inst/include
Fix "ANN.h" include to <ANN/ANN.h>

Commit:
Merge with master

Commit:
Fix Makevars.win

jefferis commented:
Thanks! Are you sure that the += in makevars.win is necessary. I can recall a warning in the past that this was a non standard gnu make extension. But perhaps that is expected on Windows.

Commit:
Revert src/Makevars

caiohamamura commented:
Yes, you're right, I reverted it back.

jefferis commented:
Thanks a lot for making all these changes. One thing I don’t understand thought. Why does the include path in makevars include /inst/ ? I always thought that the inst subdir was stripped out during the build process.

Commit:
Compile static library for linking

Commit:
Update version

Commit:
Fix ar command for unix

caiohamamura commented:
Actually everything inside/inst will be copied to where the library is
installed. You can add /inst/ to include or library path then it can be
build. If you look in the `gpuR` package they even compile the static
library inside the inst, I think we should do that for RANN2 aswell because
linking to static library works better, I will make that change.

Em qua, 19 de jun de 2019 20:40, Gregory Jefferis <notifications@github.com>
escreveu:

> Thanks a lot for making all these changes. One thing I don’t understand
> thought. Why does the include path in makevars include /inst/ ? I always
> thought that the inst subdir was stripped out during the build process.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jefferis/RANN2/pull/2?email_source=notifications&email_token=AA2H474Y35EL24SX262G2ZDP3K7WNA5CNFSM4HYPS7SKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYDSOJA#issuecomment-503785252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2H47YSQ4KBDT2LRTASGCLP3K7WNANCNFSM4HYPS7SA>
> .
>

Commit:
Forgot the LIBNAME variable in makevars

jefferis commented:
Thanks a lot for these @caiohamamura. This is great. I don't have cross-platform tests setup for this. Is there any chance that you could test against using rhub and/or win-builder on windows/Mac to make sure all is good. Thanks again.

jefferis commented:
Just to add that I have built successfully on my local Mac. I did get:
`Warning: file 'RANN2/cleanup' did not have execute permissions: corrected`
so that looks like a permissions error.

jefferis commented:
OK. I've tested on windows as well. Looks good. If you can make that cleanup permissions fix, I think all is fine. Actually one more thing, could you write a short section in the readme that would explain how to link to the package from another package?

jefferis commented:
Changed my mind. I'm going to merge now and make the executable change myself.
